### DISCUSSION TITLE
idea: rudimentary package management built into tree-sitter lib

### DISCUSSION THREAD

**Discussion by llllvvuu** [tree-sitter/tree-sitter#2662] [Ideas]
Distributing pre-generated parse tables uses much more bandwidth and hosting resources than the grammar and scanner files themselves.

Currently, the "clone and build" distribution method expects clients to have the tree-sitter CLI as well as a C toolchain. We could either remove all dependencies by distributing .wasm/.so (#730), or conversely expect a JS runtime (#2403) and greatly reduce download sizes (this would have savings increasing in the number of languages, whereas the cost is fixed). Or we could support both.

Rudimentary package management would contribute:

* A spec for parser distribution formats (analogous to .deb)
  - `langname-src-langversion-abiversion.tar.gz` - smallest
    - Contents: `grammar.js`, `scanner.o` (not technically "source", but maybe opens up to people writing in Zig/etc; must have no external dependencies), maybe `*.scm`
  - `langname-wasm-langversion-abiversion.tar.gz` - most portable without needing build toolchain
  - `langname-arch-triple-langversion-abiversion.tar.gz` - fastest without needing build toolchain
  - No format would include `parser.c` - it's even bigger than `.so` while requiring external dependency to build
  - But could a JSON-based distribution fall on the efficient frontier? In https://github.com/Julian/tree-sitter-lean, `grammar.json` is >100x smaller than `lean.so` and >500x smaller than `parser.c`
  - One challenge I could see with the `src` and `json` formats is that the `tree-sitter` runtime library is in C, but the generation stuff is in Rust?
* A common install prefix and local resolution scheme (`$TS_PARSER_PATH/$TS_MAJOR_VERSION`)
* A reference implementation / library that editors/etc can use

No ambition for a standard registry yet.

### API

```c
enum InstallFormat { SOURCE, WASM, NATIVE };
enum InstalledFormat { WASM, NATIVE };

bool ts__can_install_from_source(); // requires JS runtime and C toolchain
bool ts__can_install_from_url(); // requires some http lib
InstallStatus ts__install_from_file(
  InstallFormat format,
  char* fullname, // treesitter-python-abcdef00
  char* alias, // python
  char* filename, // path/to.tar.gz
  bool overwrite, // maybe more than 2 possible behaviors here
);
InstallStatus ts__install_from_url(
  InstallFormat format,
  char* fullname,
  char* alias,
  char* url,
  char* sha256,
  bool overwrite,
);
UninstallStatus ts__uninstall_alias(InstalledFormat format, char* alias);
UninstallStatus ts__uninstall_fullname(InstalledFormat format, char* fullname);
TSLanguage* ts__resolve_alias(InstalledFormat format, char* alias);
TSLanguage* ts__resolve_fullname(InstalledFormat format, char* fullname);
```

### CLI

```sh
tree-sitter install [--source|--wasm|--native] <FULL_NAME> <ALIAS> <URL_OR_FILENAME>
tree-sitter uninstall [--wasm|--native] <FULL_NAME_OR_ALIAS>
tree-sitter which <ALIAS | FULL_NAME>
```

### Editor/Plugin Usage
* Editor handles mapping of their internal filetype names to tree-sitter install aliases
* Editor handles running (if configured) `ts__install_from_url` whenever `ts__resolve_alias` returns `NULL`
* Editor provides `:TSInstall`/`:TSForceInstall` wrapping `ts__install_from_url`, supplying lookup table from language name to URL
* Editor uses resolved `TSLanguage` much like how they are using now. The `tree-sitter` library's resolution mechanism will only return `TSLanguage`s that it's compatible with.
* If e.g. `python` is too generic an alias for a global install, an alias like `nvim-python` or `helix-python` could be used
* Bundling parsers could be done like how Helix is doing now, but could also be done by bundling source .tar.gz and inflating on first startup (prompting QuickJS/tinycc download if no Node.js/gcc/clang found).

**Comment by clason** [comment]
> conversely expect a JS runtime (https://github.com/tree-sitter/tree-sitter/pull/2403)

Unacceptable for Neovim.

(I applaud your enthusiasm, but it would be a good idea to spend some time learning about the landscape and previous discussions on this topic -- both of which are quite complex -- first, to save everybody some time.)

The main thing is _binary_ distribution for a (very!) wide range of platforms and architectures (and hosts), ideally without requiring any external toolchain locally.  (The "editor usage" is a red herring; that part is already pretty much figured out.)

There's basically three choices:
1. require a toolchain (the current way -- this is not as horrible as you make it sound, since the `parser.c` compresses _very_ well);
2. distribute binaries for a handful of "common platform/architecture" pairs (discussed on and off, but would leave out too many platforms that have Tier 1 support in, e.g., Neovim);
3. distribute a fully portable "binary" (which these days means WASM).

**Reply by llllvvuu** [reply]
> Unacceptable for Neovim.

But what about for `nvim-treesitter`, which has higher system requirements than Neovim? (though `nvim-treesitter` might have less exact info about the tree-sitter lib version) Why is a JS dependency worse than the current system requirements?

> the landscape and previous discussions on this topic

I've seen closed issues regarding JS for a base plugin system, but I'm having trouble finding references related to specifically tree-sitter. JS is clearly a bad idea for core, but that doesn't stop periphery/plugins like coc.nvim. And having a JS-based installer is less bad than having a JS-based runtime since you can always not use the installer.

Granted, "javascript|js|node [runtime]" are not great search terms. What's the context I should be looking at?

> The main thing is binary distribution

This is exactly what I'm questioning since Tree-sitter is unlike other software in that the binaries are orders of magnitude larger than the source (even a compressed .so doesn't solve - and there may even be a survivorship bias considering there may be parsers that are straight up not distributed due to this). This makes Gentoo-style distribution not as ridiculous.

> 1. require a toolchain (the current way -- this is not as horrible as you make it sound, since the parser.c compresses very well);

Unfortunately GitHub doesn't seem to care how well it compresses when determining file size limits, and version-controlling blobs is pretty inefficient. But it sounds like we are all consider 3 > 2 > 1 already so this is a moot point.

**Reply by llllvvuu** [reply]
Also, any Neovim-specific discussions don't entirely close the case here since these tree-sitter packages are available for arbitrary consumers.

The only reason upstream guidance is relevant at all is so that parser maintainers know what formats would be helpful in general. If the guidance is to ship only binary builds because Neovim core is only willing to consume binary builds, I'm not sure how optimal that is. Especially if standardized arch-specific releases are up for consideration I'm not sure why standardized source releases would be an insane burden to add to the (hypothetical) spec.

**Reply by clason** [reply]
I only bring in Neovim as an example I am (deeply) familiar with, and arguably as currently the biggest (in terms of supported parsers and features) consumer. 

The core point here is that in the end, tree-sitter consumes binaries (`parser.so`), and these somehow have to end up somewhere the consumer can find them. This is less of an issue if there's a small list of languages you support via static compilation into whatever (say) editor binary you ship, but clearly much more of an issue if you treat these as user-installable "runtime files" -- which is crucial for Neovim, and also (planned) for Helix and Emacs. 

If your issue is only about avoiding `parser.c` as a release artifact, that's well-trodden ground and covered by the item

> Cleanup - Remove generated files from all the grammar repos in the tree-sitter org

in the roadmap issue you linked to. (This will make `tree-sitter` a mandatory build dependence of Neovim and runtime dependence of nvim-treesitter, which is not great. I hope this doesn't go as far as also removing the `grammar.js`, requiring this to be generated, too, and adding node as a requirement.)

**Reply by llllvvuu** [reply]
> user-installable "runtime files"

That's partly why I balk at the download size, downloading lean.so on a potato connection would be unpleasant. But maybe I'm being extreme here if the modal user is expected to install no more than like 10 languages + updates become less frequent.

> This will make tree-sitter a mandatory build dependence of Neovim and runtime dependence of nvim-treesitter, which is not great. I hope this doesn't go as far as also removing the grammar.js, requiring this to be generated, too, and adding node as a requirement.

Right, definitely not suggesting to just remove parser.c and *only* offer the grammar.js + scanner.c/.o version. The timeline would be (first two steps are from #730 discussion) first offer .wasm/.so version -> remove parser.c -> offer grammar.js(on?) + scanner.c version

As noted, the last thing would almost automatically exist once #730 is done (since repo directory structure is already close-to-standardized). The only minor optimizations on top of that would be stripping test files/config files/etc and/or pre-building scanner.o.

**Reply by clason** [reply]
> That's partly why I balk at the download size, downloading lean.so on a potato connection would be unpleasant. But maybe I'm being extreme here if the modal user is expected to install no more than like 10 languages + updates become less frequent.

Yes, that is the expectation. And I looked at lean; there's no reason the parser should be that large, someone™️ should try to improve the grammar to reduce state count and conflicts.

> Right, definitely not suggesting to just remove ...

What _is_ your suggestion, over what has been discussed before (and _doesn't_ require redesigning tree-sitter from scratch)?

**Reply by llllvvuu** [reply]
>> Right, definitely not suggesting to just remove ...

> What is your suggestion, over what has been discussed before (and doesn't require redesigning tree-sitter from scratch)?

That, but without the "only". i.e. guidance for distributing source-only packages (not mutually exclusive with also distributing binary packages); and `tree-sitter`-managed versioning / metadata so I can say "fetch me some compatible files" (not just "check if this file is compatible", which has been discussed before). I don't feel too strongly about having `tree-sitter`-managed parser directory (though it is repeated between editors), that part might've been clickbait in hindsight.

**Reply by llllvvuu** [reply]
I'll try to get Lean under 5 MB and if I can do that then I'm no longer worried about this.

**Reply by clason** [reply]
> just "check if this file is compatible", which has been discussed before

Without any resolution. The question of tree-sitter parser versioning has been open for some time -- again, due to the different requirements of "static" vs. "dynamic" consumers. _That_ is critical issue; once that is resolved, the rest is paint by numbers.

**Reply by llllvvuu** [reply]
>  again, due to the different requirements of "static" vs. "dynamic" consumers

Intelligence w.r.t. dynamic loading is what I had in mind with the `ts__resolve_alias` sketch. Essentially if you've installed `/ts-v2/lang-v3/{parser.so, query.scm}` and `/ts-v3/lang-v4/{parser.so, query.scm}` then `treesitter-lib` would have great context in picking the right one to load (or returning NULL if none suitable). And the remote would have to have a similar tagging scheme.

`nvim-treesitter` already hits me with downloads whenever I up/downgrade `nvim-treesitter`, although it deletes versions I've already downloaded and should in theory also correspond with some mashup of Neovim up/downgrades and `nvim-treesitter` queries, rather than just `nvim-treesitter` up/downgrades; more specifically with `tree-sitter` lib up/downgrades. `tree-sitter` lib taking over some of the logic could just know that.

(even more true if https://github.com/nvim-treesitter/nvim-treesitter/issues/3944 pans out)

But it wasn't my finest design, just meant to spark ideas.

So tl;dr the second half of the OP idea was exploring centralizing/upstreaming any of the `TSInstall` / `Language:add` logic (if not, then indeed there is nothing new over https://github.com/nvim-treesitter/nvim-treesitter/issues/3944)

**Reply by clason** [reply]
None of this is a technical issue, though; it's a strategic design issue. (Nevertheless, I would recommend to invest some time learning about the actual tech, as 
> treesitter-lib would have great context in picking the right one to load

is just not how it works, and 

> nvim-treesitter already hits me with downloads whenever I up/downgrade nvim-treesitter, although it overwrites versions I've already downloaded

also indicates that something is going wrong on your end, as that is not expected behavior. (Is it Nix? It's always Nix...))

**Reply by llllvvuu** [reply]
> also indicates that something is going wrong on your end, as that is not expected behavior. (Is it Nix? It's always Nix...))

I think it's just:

> [NOTE: This plugin is only guaranteed to work with specific versions of language parsers (as specified in the lockfile.json). When upgrading the plugin, you must make sure that all installed parsers are updated to the latest version via :TSUpdate. It is strongly recommended to automate this](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/README.md#installation)

> is just not how it works

Can the same constants not be used to search the filesystem?

https://github.com/tree-sitter/tree-sitter/blob/2cbfcbc35a7d245968f07ef5e9dee1a3ea574fee/lib/include/tree_sitter/api.h#L169-L179

https://github.com/tree-sitter/tree-sitter/blob/2cbfcbc35a7d245968f07ef5e9dee1a3ea574fee/lib/include/tree_sitter/api.h#L21-L33

("generally backwards-compatible" "earliest ABI version" :thinking:)

Like if you change the directory from `parsers/` to `parsers/v14/` and throw the `*.scm`s in there too (`parsers/v14/python-v6/*.scm`) there should be no issues right?

**Reply by clason** [reply]
This is not the channel to answer such basic questions. Feel free to ask on Matrix.

(TL;DR: This is the consumer's responsibility. And the ABI is not the issue here.)
